### PR TITLE
Tentative fix for EventState::done() data race

### PR DESCRIPTION
**Pull request by jtattermusch**
Tentative fix for #14214
Based on the TSAN listing from https://github.com/grpc/grpc/issues/14214 it is pretty clear that there's a data race between  setting are reading  the "done" flag.

I was able to reproduce the issue easily and with this change I am not able to reproduce anymore.
Repro:
```
tools/run_tests/run_tests.py -l c++ -c tsan -r 'chttp2_settings_timeout_test.*SettingsTimeout.Basic' --force_use_pollers epoll1 -n 100
```

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @yang-g

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/core/transport/chttp2/settings_timeout_test.cc [hunk 1]
```diff
@@ -179,11 +179,11 @@ class Client {
       gpr_log(GPR_INFO, "OnEventDone(): %s", grpc_error_string(error));
```
Comment (jtattermusch) [inline test/core/transport/chttp2/settings_timeout_test.cc]: It is unclear to me if OnEventDone is expected to only be called when mu_ (see PollUntilDone) is held (and therefore not doing any synchronization here is intentional). If so, there's a problem somewhere deeper in C-core stack.
If not synchronizing access to `done_` properly is just an omission, the fix is simple (= the diff in this PR).

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__262144_2_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Pull request merged**